lib/puppet/rest/client.rb

```diff
@@ -0,0 +1,82 @@
+require 'httpclient'
+
+require 'puppet'
+require 'puppet/rest/response'
+
+module Puppet::Rest
+  class Client
+    # Create a new HTTP client for querying the given API.
+    # @param [Puppet::Rest::Route] route data about the API being queried,
+    #                              including the API name and server details.
+    # @param [OpenSSL::X509::Store] ssl_store the SSL configuration for this client
+    # @param [Integer] receive_timeout how long in seconds this client will wait
+    #                  for a response after making a request
+    # @param [HTTPClient] client the third-party HTTP client wrapped by this
+    #                     class. This param is only used for testing.
+    def initialize(route,
+                   ssl_store: OpenSSL::X509::Store.new,
+                   receive_timeout: Puppet[:http_read_timeout],
+                   client: HTTPClient.new(agent_name: nil,
+                                          default_header: {
+                                            'User-Agent' => Puppet[:http_user_agent],
+                                            'X-PUPPET-VERSION' => Puppet::PUPPETVERSION
+                                          }))
+      @client = client
+
+      configure_client(ssl_store, route, receive_timeout)
+    end
+
+    # Configures the underlying HTTPClient
+    # @param [OpenSSL::X509::Store] ssl_store the SSL configuration for this client
+    # @param [Puppet::Rest::Route] route data about the API being queried
+    # @param [Integer] timeout how long to wait for a response from the server once
+    #                  a request has been made
+    def configure_client(ssl_store, route, timeout)
+      @client.tcp_keepalive = true
+      @client.connect_timeout = Puppet[:http_connect_timeout]
+      @client.receive_timeout = timeout
+      @client.base_url = route.uri
+
+      if Puppet[:http_debug]
+        @client.debug_dev = $stderr
+      end
+
+      @client.ssl_config.cert_store = ssl_store
+
+      configure_verify_mode(@client.ssl_config)
+    end
+    private :configure_client
```

Comment:
I tend to prefer this way because it's so easy to miss the private keyword, and you have to scroll around looking for it if you're trying to figure out if a method is private or not, and also easy to put your new methods in the wrong section by mistake. Guess I'm just used to languages where the visibility is part of the method signature.

That being said, I'm happy to do it the other way. I'll rearrange this.